neuralpp/symbolic/eliminator.py

```diff
@@ -0,0 +1,85 @@
+from .expression import QuantifierExpression, Context, Expression, Constant, AbelianOperation, Variable, \
+    FunctionApplication
+from .basic_expression import BasicQuantifierExpression
+from .sympy_expression import SymPyExpression
+from .interval import ClosedInterval, DottedIntervals, from_constraint
+from .constants import if_then_else
+import neuralpp.symbolic.functions as functions
+from functools import reduce
+from typing import Optional, Callable
+import operator
+
+
+def _eliminate_interval(operation: AbelianOperation, index: Variable, interval: ClosedInterval, body: Expression) \
+        -> Expression:
+    # try to solve symbolically
+    result = _symbolically_eliminate(operation, index, interval, body)
+    if result is not None:
+        return result
+
+    # TODO: enable this
+    # if isinstance(interval.lower_bound, Constant) and isinstance(interval.upper_bound, Constant):
+    #     # iterate through the interval if we can
+    #     return reduce(operation,
+    #                   map(lambda num: body.replace(interval.index, Constant(num)), iter(interval)),
+    #                   operation.identity)
+
+    return BasicQuantifierExpression(operation, index, interval.to_context(index), body)
+
+
+def _repeat_n(operation: AbelianOperation, expression: Expression, size: Expression) -> Optional[Expression]:
+    if operation.value == operator.add:
+        return expression * size
+    if operation.value == operator.mul:
+        return operator.pow(expression, size)
+    return None  # TODO: expand this
+
+
+def _symbolically_eliminate(operation: AbelianOperation, index: Variable, interval: ClosedInterval, body: Expression) \
+        -> Optional[Expression]:
+    if operation.value == operator.add:
+        if (result := Eliminator.symbolic_sum(body, index, interval)) is not None:
+            return result
+
+    if isinstance(body, Constant):
+        # repeat addition: multiplication, repeat multiplication: power
+        return _repeat_n(operation, body, interval.size)
+    return None
+
+
+def _map_leaves(conditional_intervals: Expression,
```

Comment:
This seems like a generic enough function (it doesn't seem to need to be about `DottedIntervals` in the leaves, it could just `Expression`) that could be placed in a util file.

Reply:
Yes